doc/developer/platform/reconciliation.md

```diff
@@ -0,0 +1,64 @@
+* On startup, the COMPUTE controller provides the current configuration to the COMPUTE instances.
+* COMPUTE instances can match the configuration to their own state and only apply changes.
+
+For this purpose, we interpret the commands COMPUTE controller provides to COMPUTE as an append-only log.
+Upon COMPUTE controller restart, the new configuration is/should be a suffix of the log of all previously received commands.
+
+The log can be compacted by collapsing commands that semantically supersede previous commands.
+
+Currently, the `Command` protocol does not express enough information to provide full command reconciliation.
+* Materialize initiates tables with ephemeral information.
+  For example, `show databases` reads from a table that is unconditionally initialized with `materialize`.
+  The reconciler has no good mechanism to distinguish initialization updates from others.
+  * *Current solution:* The coordinator truncates all ephemeral tables on startup using a new instruction.
+  * We could maintain a trace of all inserts and block already-inserted data.
+    This would require us to maintain a potentially long history of inserts.
+  * We cannot clear tables by advancing their compaction frontier to `[]`.
+    This would invalidate all dataflows that depend on the table.
+  * The timestamp of inserts during initialization is the current time, not 0, so we can't use this to distinguish initialization inserts.
+* The controller assumes that dataflows are rendered with a specific initial time.
+  When restarting, we do not render previously rendered dataflows, which might violate the implicit contract expected by the coordinator.
+  The reconciler tries to bring the coordinator up-to-date by replying with a correcting `FrontiersUppers` or `TimestampBindings` response.
+  There might be a brief moment in time when the two are not synchronized and the coordinator could produce invalid commands.
+* Peeks and responses are not handled correctly.
+  When reconnecting while a peek was active, we might send a response to a coordinator that doesn't know about the peek.
+  Currently, the COMPUTE controller discards peek responses it cannot match to pending peeks.
+* We need to validate if STORAGE's timestamp bindings are save to be stored in memory within COMPUTE until a coordinator picks them up.
+* The commands supplied by the COMPUTE controller should be idempotent, but in fact are not.
+  For example, when reading a Kafka topic, we create a new topic name each time we instantiate the source.
+  This is incorrect because we bind an existing identifier to a new definition.
+
+### State
+
+We assume that the mapping of `GlobalId` to an object is unique, i.e., after restarts they still name the same object.
+This allows us to maintain a small amount of state to enable command reconciling.
+
+* For each installed identifier, we maintain its upper frontier.
```

Comment:
"We" is not clear to me here. I think the two participants are the controller and the instance. We probably have the frontier on both of them, but which one are we speaking about here?